vm/src/coroutine.rs

```diff
@@ -1,55 +1,60 @@
 use crate::{
-    builtins::{PyBaseExceptionRef, PyStrRef, PyTypeRef},
+    builtins::{PyBaseExceptionRef, PyStrRef},
     common::lock::PyMutex,
     exceptions,
     frame::{ExecutionResult, FrameRef},
-    PyObjectRef, PyResult, TypeProtocol, VirtualMachine,
+    protocol::PyIterReturn,
+    IdProtocol, PyObjectRef, PyResult, TypeProtocol, VirtualMachine,
 };
 use crossbeam_utils::atomic::AtomicCell;
 
-#[derive(Debug, PartialEq, Clone, Copy)]
-pub enum Variant {
-    Gen,
-    Coroutine,
-    AsyncGen,
-}
-impl Variant {
-    fn exec_result(self, res: ExecutionResult, vm: &VirtualMachine) -> PyResult {
-        res.into_result(self == Self::AsyncGen, vm)
-    }
-    fn name(self) -> &'static str {
-        match self {
-            Self::Gen => "generator",
-            Self::Coroutine => "coroutine",
-            Self::AsyncGen => "async generator",
-        }
-    }
-    fn stop_iteration(self, vm: &VirtualMachine) -> PyTypeRef {
+impl ExecutionResult {
+    /// Turn an ExecutionResult into a PyResult that would be returned from a generator or coroutine
+    fn into_iter_return(self, vm: &VirtualMachine) -> PyIterReturn {
         match self {
-            Self::AsyncGen => vm.ctx.exceptions.stop_async_iteration.clone(),
-            _ => vm.ctx.exceptions.stop_iteration.clone(),
+            ExecutionResult::Yield(value) => PyIterReturn::Return(value),
+            ExecutionResult::Return(value) => {
+                let arg = if vm.is_none(&value) {
+                    None
+                } else {
+                    Some(value)
+                };
+                PyIterReturn::StopIteration(arg)
+            }
         }
     }
 }
 
 #[derive(Debug)]
 pub struct Coro {
     frame: FrameRef,
-    pub closed: AtomicCell<bool>,
+    pub closed: AtomicCell<bool>, // redudant?
```

Comment:
If frame was made into an option `closed` could be removed. That's closer to what CPython does.

Reply:
I replaced this comment to link to this comment